From: Gleb Natapov <gleb@redhat.com>
Date: Thu, 20 Dec 2012 16:57:42 +0200
Subject: [PATCH] KVM: emulator: drop RPL check from linearize() function
Git-commit: 3a78a4f46302bfc83602a53dfa4dcbe76a7a1f5f
Patch-mainline: v3.9-rc1
References: bnc#754583

According to Intel SDM Vol3 Section 5.5 "Privilege Levels" and 5.6
"Privilege Level Checking When Accessing Data Segments" RPL checking is
done during loading of a segment selector, not during data access. We
already do checking during segment selector loading, so drop the check
during data access. Checking RPL during data access triggers #GP if
after transition from real mode to protected mode RPL bits in a segment
selector are set.

Signed-off-by: Gleb Natapov <gleb@redhat.com>
Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
Acked-by: Bruce Rogers <brogers@suse.com>
---
 arch/x86/kvm/emulate.c |    4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

Index: b/arch/x86/kvm/emulate.c
===================================================================
--- a/arch/x86/kvm/emulate.c
+++ b/arch/x86/kvm/emulate.c
@@ -585,7 +585,7 @@ static int __linearize(struct x86_emulat
 	ulong la;
 	u32 lim;
 	u16 sel;
-	unsigned cpl, rpl;
+	unsigned cpl;
 
 	la = seg_base(ctxt, ctxt->ops, addr.seg) + addr.ea;
 	switch (ctxt->mode) {
@@ -620,8 +620,6 @@ static int __linearize(struct x86_emulat
 				goto bad;
 		}
 		cpl = ctxt->ops->cpl(ctxt);
-		rpl = sel & 3;
-		cpl = max(cpl, rpl);
 		if (!(desc.type & 8)) {
 			/* data segment */
 			if (cpl > desc.dpl)
